[obj2yaml] Add ability to dump GOFF header records

redstar opened the pull request:
Add support for the GOFF file format to obj2yaml, beginning with dumping header records.
The GOFF file format does not fit well into the ELF model, so an iterator over the
logical records and a data extractor is used to retrieve the information.

redstar committed:
[obj2yaml] Add ability to dump GOFF header records

Add support for the GOFF file format to obj2yaml, beginning with dumping header records.
The GOFF file format does not fit well into the ELF model, so an iterator over the
logical records and a data extractor is used to retrieve the information.

redstar committed:
Accidently removed an error check.

jh7370 commented:
Hi @redstar, I'm not going to have much time to look at this PR. Are you aware of #71445? I think we should land the yaml2obj side that @ysyeda has been working on before we really progress on this one.

MaskRay commented on llvm/lib/Object/GOFFObjectFile.cpp:
```diff
@@ -24,6 +25,94 @@
 
+namespace {
+// Return the type of the record.
+GOFF::RecordType getRecordType(const uint8_t *PhysicalRecord) {
```
nit: prefer `static` for functions. See https://llvm.org/docs/CodingStandards.html#anonymous-namespaces

MaskRay reviewed (commented):
.

redstar commented:
> Hi @redstar, I'm not going to have much time to look at this PR. Are you aware of https://github.com/llvm/llvm-project/pull/71445? I think we should land the yaml2obj side that @ysyeda has been working on before we really progress on this one.

Yes, I am of that PR. I'll try to develop both in parallel because that really helps with testing.

redstar committed:
Address reviewer comments and more

- Use static instead of anonymous namespace for functions
- Use Twine instead of string concatenation
- Fix clang-format error
- Move test cases to test/obj2yaml since they are for the tool

redstar committed:
Replace std::to_string with Twine

redstar commented on llvm/lib/Object/GOFFObjectFile.cpp:
```diff
@@ -24,6 +25,95 @@
+      if (RecordType != CurrRecordType) {
+        createError(object_error::parse_failed,
+                    Twine("Record ")
+                        .concat(std::to_string(RecordNum))
```
Thanks for pointing that out! I was not aware of that appraoch.
